src/twisted/web/http.py

```diff
@@ -1129,9 +1129,12 @@ def write(self, data):
             # if we don't have a content length, we send data in
             # chunked mode, so that we can support pipelining in
             # persistent connections.
+            # For no-body responses and CONNECT requests this
+            # does not apply and the header must not be sent.
             if ((version == b"HTTP/1.1") and
                 (self.responseHeaders.getRawHeaders(b'content-length') is None) and
-                self.method != b"HEAD" and self.code not in NO_BODY_CODES):
+                self.method != b"HEAD" and self.code not in NO_BODY_CODES and
+                (self.method != b"CONNECT" and self.code // 100 == 2)):
```

Comment:
Hmm... is there any other way that `transfer-encoding` headers could get set?  Possibly beyond the scope of this PR, but it seems like we should validate that somehow as well.

(Just aesthetically, I'd rather `200 <= self.code < 300` than `// 100 == 2`)

Reply:
(Also, should this just be `== 200`? What would it even mean to respond to `CONNECT` with 201 Created or 204 No Content?)